File: 06/main.js
#!/usr/bin/env node
console.log(
  require('fs')
  .readFileSync(process.env.INPUT || process.argv[2] || 'input.txt', 'utf8')
  .trim()
  .split('\n\n')
  .map(s => {
    const group = s.split('\n');
    const groupSize = group.length;
    const counts = {};
    for (const person of group) {
      for (const answer of person) {
        counts[answer] = (counts[answer] | 0) + 1;
      }
    }
    let partOne = 0;
    let partTwo = 0;
    for (const question in counts) {
      ++partOne;
      partTwo += (counts[question] === groupSize);
    }
    return [partOne, partTwo];
  })
  .reduce((sums, groupValues) => [
    sums[0] + groupValues[0],
    sums[1] + groupValues[1],
  ], [0, 0])
);

File: 06/main.cpp
#include <cstdint>
#include <iostream>
#include <fstream>


int main(int const argc, char const* const* const argv) {
  std::ifstream ifs(argc < 2 ? DEFAULT_INPUT_FILE : argv[1]);

  unsigned partOne = 0;
  unsigned partTwo = 0;
  while (ifs.good()) { // Each group.
    // Questions are all marked a-z, so a uint32_t (and ASCII-only!) is fine.
    std::uint32_t groupAnswers[26] = {0};
    unsigned groupSize = 1;

    while (ifs.good()) { // Each person.
      char const question = ifs.get();
      if (question != '\n') {
        ++groupAnswers[question - 'a'];
        continue;
      }

      if (ifs.peek() == '\n') {
        // Group complete.
        ifs.get();
        break;
      }
      if (ifs.peek() == std::ifstream::traits_type::eof()) {
        // EOF.
        break;
      }

      // Next person in group to follow.
      ++groupSize;
    }

    for (unsigned question = 0; question < 26; ++question) {
      partOne += (groupAnswers[question] != 0);
      partTwo += (groupAnswers[question] == groupSize);
    }
  }

  std::cout << "Part one: " << partOne << "\nPart two: " << partTwo
    << std::endl;
  return 0;
}
